src/Table.cpp

```diff
@@ -359,191 +433,440 @@ namespace galsim {
     //lookup and interpolate an array of function values.
     void Table::interpMany(const double* argvec, double* valvec, int N) const
     {
-        for (int k=0; k<N; k++) {
-            valvec[k] = _pimpl->lookup(argvec[k]);
-        }
+        _pimpl->interpMany(argvec, valvec, N);
     }
 
     void TableBuilder::finalize()
     {
-        _pimpl.reset(new TableImpl(&_xvec[0], &_fvec[0], _xvec.size(), _in));
+        if (_in == Table::gsinterp)
+            _makeImpl(&_xvec[0], &_fvec[0], _xvec.size(), _gsinterp);
+        else
+            _makeImpl(&_xvec[0], &_fvec[0], _xvec.size(), _in);
         _final = true;
     }
 
-    // Table2DImpl
-    class Table2D::Table2DImpl
-    {
-    public:
-        Table2DImpl(const double* xargs, const double* yargs, const double* vals,
-                    int Nx, int Ny, Table2D::interpolant in);
-
-        double lookup(double x, double y) const;
-        double lookup(double x, double y, int i) const;
-        int prelookup_i(double x) const;
-        void gradient(double x, double y, double& dfdx, double& dfdy) const;
+    // The hierarchy for Table2DImpl looks like:
+    // Table2DImpl <- ABC
+    // T2DCRTP<T> : Table2DImpl <- curiously recurring template pattern
+    // T2DLinearInterp : T2DCRTP<T2DLinearInterp>
+    // ... similar, Floor, Ceil, Nearest, Spline
+    // T2DGSInterpolant<interpolant> : T2DCRTP<T2DGSInterpolant<interpolant>> <- Use Interpolant
 
-    private:
-        Table2D::interpolant _iType;
+    class Table2D::Table2DImpl {
+    public:
+        Table2DImpl(const double* xargs, const double* yargs, const double* vals, int Nx, int Ny) :
+            _xargs(xargs, Nx), _yargs(yargs, Ny), _vals(vals), _ny(Ny) {}
+
+        virtual double lookup(double x, double y) const = 0;
+        virtual void interpMany(const double* xvec, const double* yvec, double* valvec,
+                                int N) const = 0;
+        virtual void gradient(double x, double y, double& dfdx, double& dfdy) const = 0;
+        virtual void gradientMany(const double* xvec, const double* yvec,
+                                  double* dfdxvec, double* dfdyvec, int N) const = 0;
+    protected:
         const ArgVec _xargs;
         const ArgVec _yargs;
         const double* _vals;
         const int _ny;
+    };
+
+
+    template<class T>
+    class T2DCRTP : public Table2D::Table2DImpl {
+    public:
+        using Table2D::Table2DImpl::Table2DImpl;
+
+        double lookup(double x, double y) const {
+            int i = _xargs.upperIndex(x);
+            int j = _yargs.upperIndex(y);
+            return static_cast<const T*>(this)->interp(x, y, i, j);
+        }
+
+        void interpMany(const double* xvec, const double* yvec, double* valvec, int N) const {
+            for (int k=0; k<N; k++) {
+                int i = _xargs.upperIndex(xvec[k]);
+                int j = _yargs.upperIndex(yvec[k]);
```

Comment:
Sure.  I added a `grid=True` kwarg to `__call__` and `__gradient__` that will cause the returned values to be on the outer product of `x` and `y` input arrays instead of returning something congruent to both of them.  This sped up a 100x100 interpolation by a factor of ~1.25, a 1000x1000 by ~2, and a 10000x10000 by ~4.3.  Since the number of calls to upperIndex should be going from N^2 to 2N, it seems like upperIndex must not be all that dominant to the runtime (using bilinear interpolation).